javascript/jquery-sectionsnap.js

```diff
@@ -0,0 +1,86 @@
+(function($) {
+	$.fn.sectionsnap = function( options ) {
+		var settings = $.extend({
+			'delay'				: 100 // time dilay (ms)
+			, 'selector'		: ".section" // selector
+			, 'reference'		: .9 // % of window height from which we start
+			, 'animationTime'	: 400 // animation time (snap scrolling)
+			, 'offsetTop'		: 0 // offset top (no snap before scroll reaches this position)
+			, 'offsetBottom'	: 0 // offset bottom (no snap after bottom - offsetBottom)
+		}, options);
+
+		var $wrapper = this
+		, direction = 'down'
+		, currentScrollTop = $(window).scrollTop()
+		, scrollTimer
+		, animating = false;
+
+		// check the direction
+		var updateDirection = function() {
+			if ($(window).scrollTop() >= currentScrollTop)
+				direction = 'down';
+			else
+				direction = 'up';
+			currentScrollTop = $(window).scrollTop();
+		}
+
+		// return the closest element (depending on direction)
+		var getClosestElement = function() {
+			var $list = $wrapper.find(settings.selector)
+			, wt = $(window).scrollTop()
+			, wh = $(window).height()
+			, refY = wh * settings.reference
+			, wtd = wt + refY - 1
+			, $target;
+
+			if (direction == 'down') {
+				$list.each(function() {
+					var st = $(this).position().top;
+					if ((st > wt) && (st <= wtd)) {
+						$target = $(this);
+						return false; // just to break the each loop
+					}
+				});
+			} else {
+				wtd = wt - refY + 1;
+				$list.each(function() {
+					var st = $(this).position().top;
+					if ((st < wt) && (st >= wtd)) {
+						$target = $(this);
+						return false; // just to break the each loop
+					}
+				});
+			}
+			return $target;
+		}
+
+		// snap
+		var snap = function() {
+			var $target = getClosestElement();
+			if ($target) {
+				animating = true;
+				$('html, body').animate({
+						scrollTop: ($target.offset().top)
+					}, settings.animationTime, function() {
+						window.clearTimeout(scrollTimer);
+						animating = false;
+					});
+			}
+		}
+		// on window scroll
+		var windowScroll = function() {
+			if (animating)
+				return;
+			var st = $(window).scrollTop();
+			if (st < settings.offsetTop)
+				return;
+			if (st > ($("html").height() - $(window).height() - settings.offsetBottom))
+				return;
+			updateDirection();
+			window.clearTimeout(scrollTimer);
+			scrollTimer = window.setTimeout(snap, settings.delay);
+		}
+		$(window).scroll(windowScroll);
+		return this;
+	};
+})(jQuery);
```

Comment:
Hey Eddie,

I designed the page such that each of the main sections would fit exactly into the "viewing frame".  The links on the nav bar use some JS that will auto-scroll and snap to the appropriate section.


The bit of JS you're referring to was supposed to auto-snap to the nearest section if the user manually was scrolling through the page.  But I wasn't able to get it to work properly before the assignment was due, and not referencing it in the assignment.  Planning to revisit it when I get some time.


I can't take credit for any of the JS, I found them both online.  Looking forward to the intermediate class where hopefully I'll be able to code things like this myself.


________________________________
From: Eddie Garcia <notifications@github.com>
Sent: Saturday, August 27, 2016 9:56 AM
To: AustinCodingAcademy/intro-workbook
Cc: crackoner; Author
Subject: Re: [AustinCodingAcademy/intro-workbook] austin.intro.2.wong.michael (#97)


In javascript/jquery-sectionsnap.js<https://github.com/AustinCodingAcademy/intro-workbook/pull/97#discussion_r76516588>:

> +             var windowScroll = function() {
> +                     if (animating)
> +                             return;
> +                     var st = $(window).scrollTop();
> +                     if (st < settings.offsetTop)
> +                             return;
> +                     if (st > ($("html").height() - $(window).height() - settings.offsetBottom))
> +                             return;
> +                     updateDirection();
> +                     window.clearTimeout(scrollTimer);
> +                     scrollTimer = window.setTimeout(snap, settings.delay);
> +             }
> +             $(window).scroll(windowScroll);
> +             return this;
> +     };
> +})(jQuery);


Damn, Mike! That's rad! What does this do exactly?

-
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/AustinCodingAcademy/intro-workbook/pull/97/files/c87450374bfc81166d27f048393f71344d5fad30..5d7e27ffcecc9ecef87f40ed636c78734790f2c6#r76516588>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ATsJVWrxpmE8z2Xz02JT7Nv0MD19jnN7ks5qkEIQgaJpZM4JbKwS>.